torch/csrc/api/src/nn/modules/transformer.cpp

```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+      .dropout(options.dropout())
+      .activation(options.activation()), options.num_encoder_layers())
+        .norm(AnyModule(norm)));
+
+    this->encoder = AnyModule(trans_encoder);
+  }
+  else {
+    this->encoder = options.custom_encoder().clone();
+  }
+  this->register_module("encoder", this->encoder.ptr());
+
+  // set up decoder
+  if (options.custom_decoder().is_empty()) {
+
+    LayerNorm norm(LayerNormOptions({options.d_model()}));
+    TransformerDecoder trans_decoder(TransformerDecoderOptions(
+      TransformerDecoderLayerOptions(options.d_model(), options.nhead())
+      .dim_feedforward(options.dim_feedforward())
+      .dropout(options.dropout())
+      .activation(options.activation()), options.num_decoder_layers())
+        .norm(AnyModule(norm)));
+
+    this->decoder = AnyModule(trans_decoder);
+  }
+  else {
+    this->decoder = options.custom_decoder().clone();
+  }
+  this->register_module("decoder", this->decoder.ptr());
+
+  reset_parameters();
+}
+
+
+void TransformerImpl::reset_parameters() {
+  auto parameters = this->parameters();
+  for (auto& param : parameters) {
+    if (param.dim() > 1) {
+      torch::nn::init::xavier_uniform_(param);
+    }
+  }
+}
+
+
+Tensor TransformerImpl::forward(
+  const Tensor& src,
+  const Tensor& tgt,
+  const Tensor& src_mask,
+  const Tensor& tgt_mask,
+  const Tensor& memory_mask,
+  const Tensor& src_key_padding_mask,
+  const Tensor& tgt_key_padding_mask,
+  const Tensor& memory_key_padding_mask) {
+
+  TORCH_CHECK(src.dim() == 3 && tgt.dim() == 3,
+    "src and tgt should have 3 dimensions, but got ", src.dim(), " and ", tgt.dim());
+
+  TORCH_CHECK(src.size(1) == tgt.size(1),
+    "src and tgt should have same size at dimension 1, but got ", src.size(1), " and ", tgt.size(1));
+
+  TORCH_CHECK(src.size(2) == options.d_model() && tgt.size(2) == options.d_model(),
+    "src and tgt should have a same size as d_model at dimension 2, but got ",
+    src.size(2), " and ", tgt.size(2), " while d_model is ", options.d_model());
+
+  Tensor memory = this->encoder.forward<Tensor>(src, src_mask, src_key_padding_mask);
+  Tensor output = this->decoder.forward<Tensor>(
+    tgt, memory, tgt_mask, memory_mask, tgt_key_padding_mask, memory_key_padding_mask);
+
+  return output;
+}
+
+Tensor TransformerImpl::generate_square_subsequent_mask(int64_t sz) {
+  // Treat 0 dim valid here
+  TORCH_CHECK(sz >= 0,
+    "Input size must be non-negative to genearte a valid square subsequent mask, but got ", sz);
+
+  Tensor mask = (torch::triu(torch::ones({sz, sz})) == 1).transpose(0, 1).to(torch::kFloat32);
+
+  // check IEEE754 support here since -inf is not guaranteed to be valid on non IEEE754 platform
+  if (std::numeric_limits<float>::is_iec559) {
+    mask = mask.masked_fill(mask == 0, -std::numeric_limits<float>::infinity()).masked_fill(mask == 1, 0.f);
+  }
+  // if IEEE754 is not supported, we use the smallest float number in current platform
```

Comment:
This seems fine, but does the Python API have the same behavior?

Reply:
Python doesn't have this issue. 
I checked with @zhangguanheng66, using the smallest float number in this situation is fine.